### PR TITLE
Implement anti-spam mechanism for proposals in gov

### PR DESCRIPTION
**Pull request by kamuikatsurgi**
# Description

ref https://polygon.atlassian.net/browse/POS-2755

# Changes

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] Changes only for a subset of nodes

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Nodes audience

In case this PR includes changes that must be applied only to a subset of nodes, please specify how you handled it (e.g. by adding a flag with a default value...)

# Checklist

- [ ] I have added at least 2 reviewer or the whole pos-v1 team
- [x] I have added sufficient documentation in code
- [x] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply

# Cross repository changes

- [ ] This PR requires changes to bor 
  - In case link the PR here:
- [ ] This PR requires changes to heimdall-v2 
  - In case link the PR here:
- [ ] This PR requires changes to matic-cli
  - In case link the PR here:

## Testing

- [ ] I have added unit tests
- [ ] I have added tests to CI
- [ ] I have tested this code manually on local environment
- [ ] I have tested this code manually on remote devnet using express-cli
- [ ] I have tested this code manually on amoy/mumbai
- [ ] I have created new e2e tests into express-cli

### Manual tests

Please complete this section with the steps you performed if you ran manual tests for this functionality, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: x/gov/keeper/deposit.go [hunk 1]
```diff
@@ -285,6 +286,81 @@ func (keeper Keeper) RefundAndDeleteDeposits(ctx context.Context, proposalID uin
...
+	deposits, err := keeper.GetDeposits(ctx, proposalID)
+	if err != nil {
+		return err
+	}
+
+	var totalDeposits sdk.Coins
+	var amountPerValidator sdk.Coins
+
+	for _, deposit := range deposits {
+		depositerAddress, err := keeper.authKeeper.AddressCodec().StringToBytes(deposit.Depositor)
```
Comment (marcello33) [inline x/gov/keeper/deposit.go]: Let's rename this to `depositorAddress`

File: x/gov/keeper/deposit.go [hunk 2]
```diff
@@ -285,6 +286,81 @@ func (keeper Keeper) RefundAndDeleteDeposits(ctx context.Context, proposalID uin
...
+		if err != nil {
+			return err
+		}
+
+		totalDeposits = totalDeposits.Add(deposit.Amount...)
```
Comment (marcello33) [inline x/gov/keeper/deposit.go:327]: here `totalDeposits.Add` might encur in NPE if `totalDeposits` is `nil`. Maybe add a check?

File: x/gov/keeper/deposit.go [hunk 3]
```diff
@@ -285,6 +286,81 @@ func (keeper Keeper) RefundAndDeleteDeposits(ctx context.Context, proposalID uin
...
+
+		for _, coin := range deposit.Amount {
+			amountPerValidatorPerCoin := sdkmath.LegacyNewDecFromInt(coin.Amount).QuoInt64(int64(numValidators)).TruncateInt()
+			amountPerValidator = amountPerValidator.Add(
```
Comment (marcello33) [inline x/gov/keeper/deposit.go:331]: same as before, `nil` check

File: x/gov/keeper/deposit.go [hunk 4]
```diff
@@ -285,6 +286,81 @@ func (keeper Keeper) RefundAndDeleteDeposits(ctx context.Context, proposalID uin
...
+
+	for _, validator := range validatorAddresses {
+		err = keeper.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, validator, amountPerValidator)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Calculate any remaining amount due to truncating issues
+	usedAmount := amountPerValidator.MulInt(sdkmath.NewInt(int64(numValidators)))
```
Comment (marcello33) [inline x/gov/keeper/deposit.go:355]: same here, nil check

File: x/gov/keeper/deposit.go [hunk 5]
```diff
@@ -285,6 +286,81 @@ func (keeper Keeper) RefundAndDeleteDeposits(ctx context.Context, proposalID uin
...
+	remainingAmount, _ := totalDeposits.SafeSub(usedAmount...)
```
Comment (marcello33) [inline x/gov/keeper/deposit.go]: same here, nil check

File: x/gov/keeper/deposit.go [hunk 6]
```diff
@@ -285,6 +286,81 @@ func (keeper Keeper) RefundAndDeleteDeposits(ctx context.Context, proposalID uin
 	})
 }
 
+// DistributeAndDeleteDeposits distributes the deposits evenly among all the active validators and deletes the deposits on a specific proposal.
+func (keeper Keeper) DistributeAndDeleteDeposits(ctx context.Context, proposalID uint64) error {
```
Comment (marcello33) [inline x/gov/keeper/deposit.go:292]: I would add some logs in this function, to be sure we print errors when returning them

File: x/gov/keeper/deposit.go [hunk 7]
```diff
@@ -9,6 +9,7 @@ import (
 	"cosmossdk.io/errors"
 	sdkmath "cosmossdk.io/math"
 
+	stakeTypes "github.com/0xPolygon/heimdall-v2/x/stake/types"
```
Comment (marcello33) [inline x/gov/keeper/deposit.go]: let's move this to the end of imports (or linter would complain)

File: x/gov/keeper/deposit.go (hunk 1 above)
Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go]: resolved

File: x/gov/keeper/deposit.go (hunk 2 above)
Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go:330]: skiped

File: x/gov/keeper/deposit.go (hunk 3 above)
Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go:331]: skipped

File: x/gov/keeper/deposit.go (hunk 2 above)
Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go:327]: skipped

File: x/gov/keeper/deposit.go (hunk 4 above)
Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go:355]: skipped

File: x/gov/keeper/deposit.go (hunk 5 above)
Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go]: skipped

File: x/gov/keeper/deposit.go (hunk 6 above)
Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go:292]: added

File: x/gov/keeper/deposit.go (hunk 7 above)
Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go]: done

**Review by marcello33 — APPROVED** [review]

File: x/gov/keeper/deposit.go [hunk 8]
```diff
@@ -285,6 +287,91 @@ func (keeper Keeper) RefundAndDeleteDeposits(ctx context.Context, proposalID uin
 	})
 }
 
+// DistributeAndDeleteDeposits distributes the deposits evenly among all the active validators and deletes the deposits on a specific proposal.
+func (keeper Keeper) DistributeAndDeleteDeposits(ctx context.Context, proposalID uint64) error {
+	var validatorAddresses []sdk.AccAddress
+
+	err := keeper.sk.IterateCurrentValidatorsAndApplyFn(ctx, func(validator stakeTypes.Validator) bool {
```
Comment (avalkov) [inline x/gov/keeper/deposit.go:295]: One thing I notice about `IterateCurrentValidatorsAndApplyFn` that the functor it accepts doesn't return error. So error can happen inside, it will trigger the iteration to stop at some point without iterating over all of the validators, potentially creating inconsistent state. 

I was working before on fixing `IterateCurrentValidatorsAndApplyFn` but haven't noticed that. Its used in at least one more place in the `gov` module.

Comment (marcello33) [reply x/gov/keeper/deposit.go:295]: Good point, I think it's worth checking it, for this function and all the other "AndApplyFn" we have in v2 and cosmos

File: x/gov/keeper/deposit.go [hunk 9]
```diff
@@ -285,6 +287,91 @@ func (keeper Keeper) RefundAndDeleteDeposits(ctx context.Context, proposalID uin
...
+		validatorAddresses = append(validatorAddresses, validatorAddr)
+		return false
+	})
+	if err != nil {
+		keeper.Logger(ctx).Error("Error iterating over validators", "error", err)
+		return err
+	}
+
+	numValidators := len(validatorAddresses)
+	if numValidators == 0 {
```
Comment (avalkov) [inline x/gov/keeper/deposit.go]: I think this check is not needed, how you can have this block produced, if there were no validators?

File: x/gov/keeper/deposit.go (hunk 8 above)
Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go:295]: It's used in the `Tally` function in `gov/keeper/tally.go`. It returns `true` if `err != nil` and `false` otherwise. I have done the same here.

File: x/gov/keeper/deposit.go (hunk 9 above)
Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go]: Just a sanity check I thought, will remove it.

File: x/gov/keeper/deposit.go [hunk 10]
```diff
@@ -285,6 +287,91 @@ func (keeper Keeper) RefundAndDeleteDeposits(ctx context.Context, proposalID uin
...
+		}
+	}
+
+	// Calculate any remaining amount due to truncating issues
+	usedAmount := amountPerValidator.MulInt(sdkmath.NewInt(int64(numValidators)))
+	remainingAmount, _ := totalDeposits.SafeSub(usedAmount...)
+
+	if !remainingAmount.IsZero() {
+		// Send remaining amount to the first validator
+		err = keeper.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, validatorAddresses[0], remainingAmount)
```
Comment (avalkov) [inline x/gov/keeper/deposit.go]: I think its not big deal because there wont be so many gov proposals and remaining will be little, but wouldn't it be better strategy to send to the FeeCollector so it will distribute it to the block proposer, instead of sending to first validator which could be same address every time.

File: x/gov/keeper/deposit.go (hunk 8 above)
Comment (avalkov) [reply x/gov/keeper/deposit.go:295]: Yea, thats how we used it. But its good to improve such iteration functions to return errors from the functor.

File: x/gov/keeper/deposit.go [hunk 11]
```diff
@@ -285,6 +287,91 @@ func (keeper Keeper) RefundAndDeleteDeposits(ctx context.Context, proposalID uin
 	})
 }
 
+// DistributeAndDeleteDeposits distributes the deposits evenly among all the active validators and deletes the deposits on a specific proposal.
+func (keeper Keeper) DistributeAndDeleteDeposits(ctx context.Context, proposalID uint64) error {
```
Comment (avalkov) [inline x/gov/keeper/deposit.go:292]: Can we have some unit tests that cover the whole execution of this function including edge cases like distributing the remaining amount.

File: x/gov/keeper/deposit.go [hunk 12]
```diff
@@ -285,6 +287,91 @@ func (keeper Keeper) RefundAndDeleteDeposits(ctx context.Context, proposalID uin
...
+		err = keeper.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, validator, amountPerValidator)
+		if err != nil {
+			keeper.Logger(ctx).Error("Failed to send coins to validator", "error", err)
+			return err
+		}
+	}
+
+	// Calculate any remaining amount due to truncating issues
+	usedAmount := amountPerValidator.MulInt(sdkmath.NewInt(int64(numValidators)))
+	remainingAmount, _ := totalDeposits.SafeSub(usedAmount...)
```
Comment (avalkov) [inline x/gov/keeper/deposit.go]: I think its good idea to also check for the error returned by SafeSub

Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go]: > I think its good idea to also check for the error returned by SafeSub

It's not returning an err, but a bool if there is a negative amount used.

File: x/gov/keeper/deposit.go (hunk 11 above)
Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go:292]: adding now.

File: x/gov/keeper/deposit.go (hunk 12 above)
Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go]: Which I don't think will be the case here.

File: x/gov/keeper/deposit.go (hunk 10 above)
Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go]: Then there's again the problem of not being able to send across module from `gov` and also I recall somewhere that `feeCollector` was banned from receiving funds from modules. I don't think there's a big difference in the amount here. Check out the test case for the difference in amount. 

Do you think we can introduce a random number and pick a random validator to send the remaining amount?

Comment (avalkov) [reply x/gov/keeper/deposit.go]: You can introduce a random number but it must be derived in a sense that each validator produces the same number, you can use rand.Seed with lets say validators count and total deposit amount maybe.

File: x/gov/keeper/deposit.go (hunk 12 above)
Comment (avalkov) [reply x/gov/keeper/deposit.go]: Just to be on the safe side, I think its not a problem to add the check.

Comment (avalkov) [reply x/gov/keeper/deposit.go]: Because this is the code now, but somebody else may do changes on it after few months and introduce bug, that why such changes can save lives.

Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go]: okay.

File: x/gov/keeper/deposit.go (hunk 10 above)
Comment (kamuikatsurgi) [reply x/gov/keeper/deposit.go]: Done.

**Review by avalkov — APPROVED** [review]
